### ISSUE TITLE
Examples of AADL with HAMR/BriefCASE-virtualization components?

### ISSUE THREAD

**Issue by jdg-11** [sireum/case-env#6]
Hi,

I'm curious if there are any available examples of using HAMR with the BriefCASE virtualization transform.  I was trying to follow:
 
    https://github.com/loonwerks/CASE/blob/master/CASE_Simple_Example_V4/BriefCASE%20Workflow%20-Simple%20UAV%20Example.pdf

But some things seem to have changed.  For example, `HAMR::Component_Type => VIRTUAL_MACHINE;` seems to not exist and causes issues in for example [CASE repo SW.aadl](https://github.com/loonwerks/CASE/tree/master/CASE_Simple_Example_V4/Hardened-VM/SW.aadl).

When I try to change things enough so that HAMR code gen works, I can get to the stage where the HAMR code gen works, the transpile slang step works, but I get linker errors in the main build.  

I'm having a little trouble understanding what to try.  Could anyone point me towards a simple aadl (or how to create one) with Linux VM components, and a general outline of what to do to get an seL4 system that builds it (e.g. run HAMR code gen then run the  transpile, setup, and build/run scripts)?

**Comment by jasonbelt** [comment]
Yes, we dropped support for the HAMR::Component_Type method of identifying VM bound processes and instead adopted the method used by the other CASE tools (ie. a process bound to a virtual processor which is bound to an actual processor).  The model your referring to is compliant, but removing the references to HAMR::Component_Type will cause a different issue if your using a current case-env.

In late Dec 21 UNSW updated their docker scripts to use Debian 11 rather than 10 which caused case-env builds to break.  @robby-phd resolved the issues and one of the required fixes was to upgrade the seL4 and CAmkES dependencies.  The new version of CAmkES folds in what had been a separate project, camkes-arm-vm, and enough changes were made such that any model containing VMs break you try to build them via CAmkES.  It appears HAMR could be modified to adapt to those changes for models containing a single VM, but the support for multiple VMs that we had previously wouldn't really be feasible until QEMU support is added -- https://github.com/seL4/camkes-vm-examples#arm-features

So currently the only way to get the model your referring to go through HAMR codegen and also CAmkES/seL4/QEMU is to use an earlier version of the case-env.  We do have an OVA for this purpose that we use in our reg testing and we could make that publicly available if needed.  Ideally though it would be best if the OVA that was used when the PDF was written were still available (i.e. it would contain the exact version of the CASE tools used at the time).  @iamundson might know if that was archived.

**Comment by ANogin** [comment]
@jasonbelt thanks for the explanation, but note that we are targeting x86 rather than ARM. Does that affect things in any significant way?

**Comment by jdg-11** [comment]
@jasonbelt : Gavin Holland tracked down the source of the linker error mentioned above.  It is a bug in seL4 that is patched in recent versions of seL4, but not the version pulled by HAMR.  

Here are steps used to create the original error.

1. Load the producer-consumer tutorial project in aadl
2. Apply the virtualization transform to either the producer or consumer
3. Adjust the aadl to fill out domain scheduling and timing information to be compliant with HAMR
4. Generate the seL4 backend with HAMR
5. Run the slang transpile script
6. Run the setup-camkes-arm-vm.sh in seL4 tooling docker container set to leave camkes in /host/CASE
7. Run the run-camkes.sh script in the seL4 tooling docker container with `-o "-DUSE_PRECONFIGURED_ROOTFS=ON -s -c /host/CASE"

The build eventually fails on multiple definitions of dev_vgic_dist.  The problem is due to https://github.com/seL4/seL4_projects_libs/commit/b0573e4896f87bb1b898b39c6ee044fbc2221095.  Indeed a small patch inserted between steps 6 and 7 allows the build to finish.  In the above, /host/CASE was used as the camkes-arm-vm directory, and from that directory:

Change projects/seL4_projects_libs/libsel4vm/src/arch/arm/vgic/vgic.h line 15 by preprending the `extern` keyword (as described in the fix for seL4)

**Comment by jasonbelt** [comment]
@jdg-11 thanks for the information.  Since our last meeting I found that the camkes-vm documentation stating that multiple VMs is not supported for QEMU ARM is incorrect.  Even then getting the devices setup correctly was not straightforward as it can be difficult to figure out what the cause of errors are (related email [thread](https://lists.sel4.systems/hyperkitty/list/devel@sel4.systems/message/E47OZC5GV5DU6WJI3XYAGPVFMRSUUR4F/)).

I've modified HAMR codegen to produce camkes-vm compliant projects for models that contain VMs.  The new version is installed in the published case-env.ova (links available from [Sireum Kekinian's](https://github.com/sireum/kekinian) and [case-env's](https://github.com/sireum/case-env) readme).  If you installed Sireum manually then you can update it by doing a ``git pull --recurse-submodules`` and then ``$SIREUM_HOME/bin/build.cmd``

Note that the HAMR generated run-camkes.sh script now expects the [case-vm-examples](https://github.com/seL4/camkes-vm-examples) project to be used to build projects that contain VMs.  This project is already included in the case-env.ova and would need to be installed if you're using the seL4/CAmkES docker containers.

I copied the HAMR examples from the tool-assessment-4 repo and regenerated everything.  It's temporarily available [here](https://github.com/jasonbelt/hamr-case-examples/tree/master/case/tool-assessment-4) -- we have a meeting today with Collins and will discuss whether the loonwerks version should be updated.

I tried a few of the examples in an seL4 docker container and they worked as expected.  For any of the examples you'll at least need to transpile the checked in code -- I did this using the sireum available from my host computer rather than installing sireum inside of the container.  For example, for the 'receiver in a vm' project [here](https://github.com/jasonbelt/hamr-case-examples/tree/master/case/tool-assessment-4/vm/test_event_data_port_periodic_domains_VM/receiver_vm) I did the following in my host

```
hamr-case-examples/case/tool-assessment-4/vm/test_event_data_port_periodic_domains_VM/receiver_vm/hamr/slang/bin/transpile-sel4.cmd
```

and then inside the container I did

```
hamr-case-examples/case/tool-assessment-4/vm/test_event_data_port_periodic_domains_VM/receiver_vm/hamr/camkes/bin/run-camkes.sh -c /host/camkes-vm-examples -s
```
